src/lib/device-online-state.ts

```diff
@@ -0,0 +1,317 @@
+								this.deviceOffline(uuid);
+								break;
+							default:
+								captureException(
+									new Error(
+										`An unexpected value was encountered for the target device state: ${nextState}`,
+									),
+								);
+						}
+					}).finally(() =>
+						this.rsmq.deleteMessageAsync({
+							qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+							id,
+						}),
+					);
+				} else {
+					// no message! this is thrown to restart the consume process...
+					return Bluebird.delay(1000).thenThrow(
+						new NoMessagesToConsumeSignal(),
+					);
+				}
+			})
+			.catch(err => {
+				if (err instanceof NoMessagesToConsumeSignal) {
+					return;
+				}
+
+				captureException(
+					err,
+					'DeviceStateManager ==> Error when consuming queue',
+				);
+			})
+			.then(() => this.consume());
+	}
+
+	private scheduleChangeOfStateForDevice(
+		uuid: string,
+		currentState: DeviceOnlineStates,
+		nextState: DeviceOnlineStates,
+		delay: number, // in seconds
+	) {
+		// remove the old queued state...
+		return this.redis
+			.getAsync(`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`)
+			.then(value => {
+				const { id } = JSON.parse(value) as { id: string };
+
+				if (id) {
+					return this.rsmq.deleteMessageAsync({
+						qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+						id,
+					});
+				}
+
+				return 1;
+			})
+			.catch(noop) // ignore errors when deleting the old queued state and queue the new one...
+			.then(() =>
+				this.rsmq.sendMessageAsync({
+					qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+					message: JSON.stringify({
+						uuid,
+						nextState,
+					}),
+					delay,
+				}),
+			)
+			.then(id =>
+				this.redis
+					.setAsync(
+						`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`,
+						JSON.stringify({
+							id,
+							currentState,
+						}),
+					)
+					.then(() =>
+						this.redis.expireAsync(
+							`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`,
+							delay + 5, // add a small 5 second buffer to allow some wiggle room on top of the timeout window...
+						),
+					),
+			);
+	}
+
+	public captureEventFor(uuid: string, timeoutSeconds: number) {
+		// see if we already have a queued state for this device...
+		return (
+			this.redis
+				.getAsync(`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`)
+				.then(value => {
+					const { id, currentState } = JSON.parse(value);
+
+					if (!id || DeviceOnlineStates.Online !== currentState) {
+						this.deviceOnline(uuid);
+					}
+				})
+				.catch(noop)
+				// record the activity...
+				.then(() =>
+					this.scheduleChangeOfStateForDevice(
+						uuid,
+						DeviceOnlineStates.Online,
+						DeviceOnlineStates.Timeout,
+						timeoutSeconds,
+					),
+				)
+		);
+	}
+}
+
+export const manager = new DeviceOnlineStateManager();
+
+// register the device state change handler which updates the device model...
+manager.on('change', (uuid: string, newState: DeviceOnlineStates) => {
```

Comment:
This is no big deal and it will work fine but it feels a bit odd for an object to register with itself in order to listen to its own events and it isn't all self-contained. What is your concern?

Reply:
No concerns; the original idea was to have this as an event you could register any code onto, so it would separate the function of the state manager from the code updating the API models. If this isn't something we desire, then moving the code into the main state manager class is fine.

Reply:
It's not important, ignore my comment.